### PR TITLE
Billings

### PR DESCRIPTION
**Pull request by ajgo**
Added repo, service, model, form for Billing
List of billings with statistics in event view
Generating pdf raport for billings

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -64,6 +64,28 @@
 			<version>${org.mapstruct.version}</version>
 		</dependency>
 
+		<dependency>
+			<groupId>org.openjdk.jmh</groupId>
```
Comment (adamzareba) [inline pom.xml]: do you use it in your code? if yes, then move version to properties section in pom.xml

File: pom.xml [hunk 2]
```diff
@@ -64,6 +64,28 @@
...
+			<artifactId>jmh-core</artifactId>
+			<version>1.19</version>
+		</dependency>
+		<dependency>
+			<groupId>org.openjdk.jmh</groupId>
```
Comment (adamzareba) [inline pom.xml]: do you use it in your code? if yes, then move version to properties section in pom.xml

File: pom.xml [hunk 3]
```diff
@@ -64,6 +64,28 @@
...
+			<artifactId>jmh-generator-annprocess</artifactId>
+			<version>1.19</version>
+		</dependency>
+		<dependency>
+			<groupId>com.thoughtworks.xstream</groupId>
+			<artifactId>xstream</artifactId>
+			<version>1.4.11.1</version>
```
Comment (adamzareba) [inline pom.xml]: move version to properties section in pom.xml

File: src/main/java/com/example/event_manager/controller/EventController.java [hunk 4]
```diff
@@ -18,21 +27,40 @@
...
   private final PersonService personService;
   private final EventService eventService;
   private final TaskStatusService taskStatusService;
+  private final BillingService billingService;
+
+  @GetMapping(value = "billingsRaport", produces = "application/pdf")
+  public void Billings(final HttpServletResponse response, @RequestParam final Long id)
+      throws TransformerException, IOException, FOPException {
+    final String pathToFile = eventService.pathToGeneratatedBillingsRaportForEvent(id);
+    try {
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/controller/EventController.java]: use try with resource 
https://www.baeldung.com/java-try-with-resources

File: src/main/java/com/example/event_manager/controller/EventController.java [hunk 5]
```diff
@@ -18,21 +27,40 @@
...
+      final InputStream is = new FileInputStream(pathToFile);
+      org.apache.commons.io.IOUtils.copy(is, response.getOutputStream());
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/controller/EventController.java]: remove package name before class

File: src/main/java/com/example/event_manager/utils/raport/BillingRaport.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  public String convertXmlToPdfAndSaveOnDisc()
+      throws IOException, FOPException, TransformerException {
+    generateXmlForBillingRaportSchema();
+    final File xsltFile = new File(
+        pathToBillingInStatic +"newStyle.xsl");
+    final StreamSource xmlSource = new StreamSource(new StringReader(billingsXml));
+    final FopFactory fopFactory = FopFactory.newInstance(new File(".").toURI());
+    final FOUserAgent foUserAgent = fopFactory.newFOUserAgent();
+    final OutputStream out;
+    out = new java.io.FileOutputStream(pathToBillingInStatic +"billings.pdf");
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/raport/BillingRaport.java]: redundant package name

File: src/main/java/com/example/event_manager/utils/raport/BillingRaport.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    xStream.addImplicitCollection(BillingRaportSchema.class, "billings");
+    billingsXml = xStream.toXML(billingRaportSchema);
+
+  }
+
+  public String convertXmlToPdfAndSaveOnDisc()
+      throws IOException, FOPException, TransformerException {
+    generateXmlForBillingRaportSchema();
+    final File xsltFile = new File(
+        pathToBillingInStatic +"newStyle.xsl");
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/raport/BillingRaport.java]: reformat code + move "newStyle.xsl" to instance/static variable

File: src/main/java/com/example/event_manager/utils/raport/BillingRaport.java (hunk 6 above)
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/raport/BillingRaport.java]: reformat code + move "billings.pdf" to instance/static variable

File: src/main/java/com/example/event_manager/utils/raport/BillingRaport.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    final FOUserAgent foUserAgent = fopFactory.newFOUserAgent();
+    final OutputStream out;
+    out = new java.io.FileOutputStream(pathToBillingInStatic +"billings.pdf");
+    try {
+      final Fop fop = fopFactory.newFop(MimeConstants.MIME_PDF, foUserAgent, out);
+      final TransformerFactory factory = TransformerFactory.newInstance();
+      final Transformer transformer = factory.newTransformer(new StreamSource(xsltFile));
+      final Result res = new SAXResult(fop.getDefaultHandler());
+      transformer.transform(xmlSource, res);
+      return pathToBillingInStatic +"billings.pdf";
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/raport/BillingRaport.java]: billings.pdf to instance/static variable

File: src/main/java/com/example/event_manager/utils/raport/BillingRaport.java (hunk 6 above)
Comment (ajgo) [reply src/main/java/com/example/event_manager/utils/raport/BillingRaport.java]: contenet from raport package moved to utils

File: src/main/java/com/example/event_manager/utils/BillingRaport.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.fop.apps.FopFactory;
+import org.apache.fop.apps.MimeConstants;
+
+public class BillingRaport {
+
+
+  private final String pathToBillingInStatic = "src/main/resources/static/billing/";
+  private final BillingRaportSchema billingRaportSchema;
+  private String billingsXml;
+  private String styleFileName = "newStyle.xsl";
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/BillingRaport.java]: + final

File: src/main/java/com/example/event_manager/utils/BillingRaport.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  private String pdfFileName = "billings.pdf";
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/BillingRaport.java]: + final

File: src/main/java/com/example/event_manager/utils/BillingRaport.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+  public String convertXmlToPdfAndSaveOnDisc()
+      throws IOException, FOPException, TransformerException {
+    generateXmlForBillingRaportSchema();
+    final File xsltFile = new File(
+        pathToBillingInStatic + styleFileName);
+    final StreamSource xmlSource = new StreamSource(new StringReader(billingsXml));
+    final FopFactory fopFactory = FopFactory.newInstance(new File(".").toURI());
+    final FOUserAgent foUserAgent = fopFactory.newFOUserAgent();
+    final OutputStream out = new java.io.FileOutputStream(pathToBillingInStatic + pdfFileName);
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/BillingRaport.java]: remove package name

File: src/main/java/com/example/event_manager/controller/EventController.java [hunk 12]
```diff
@@ -18,21 +24,30 @@
...
 @AllArgsConstructor()
+@RequestMapping("events")
 public class EventController {
 
   private final PersonService personService;
   private final EventService eventService;
   private final TaskStatusService taskStatusService;
 
+  @GetMapping(value = "billingsRaport", produces = "application/pdf")
+  public ResponseEntity<byte[]> Billings(@RequestParam final Long id)
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/controller/EventController.java]: minor issue but java convention for method names is to start with lowercase character, so please rename it to "billings"

File: .gitignore [hunk 13]
```diff
@@ -3,7 +3,7 @@ target/
 !.mvn/wrapper/maven-wrapper.jar
 !**/src/main/**
 !**/src/test/**
-
+src/main/resources/static/billing/billings.pdf
```
Comment (adamzareba) [inline .gitignore]: can we remove it now?

File: src/main/java/com/example/event_manager/controller/EventController.java [hunk 14]
```diff
@@ -72,7 +87,7 @@ public String editTaskStatus(final Model model, @RequestParam final Long taskId)
   @PostMapping(value = "editTask/save")
   public String saveEditedTask(@ModelAttribute(value = "task") final TaskStatusForm taskStatus) {
     taskStatusService.update((taskStatus));
-    return "redirect:/events/details?id=" + Long.valueOf(1);
+    return "redirect:/events/details?id=" + 1L;
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/controller/EventController.java]: what is this magic number 1l?

File: src/main/java/com/example/event_manager/utils/BillingRaport.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    xStream.alias("billing", BillingForm.class);
+    xStream.addImplicitCollection(BillingRaportSchema.class, "billings");
+    billingsXml = xStream.toXML(billingRaportSchema);
+
+  }
+
+  public byte[] convertBillingRaportToByteStream()
+      throws IOException, FOPException, TransformerException {
+    generateXmlForBillingRaportSchema();
+    String styleXslFile = "src/main/resources/static/billing/newStyle.xsl";
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/BillingRaport.java]: move it to static final variable

File: src/main/java/com/example/event_manager/utils/BillingRaport.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.sax.SAXResult;
+import javax.xml.transform.stream.StreamSource;
+import org.apache.fop.apps.FOPException;
+import org.apache.fop.apps.FOUserAgent;
+import org.apache.fop.apps.Fop;
+import org.apache.fop.apps.FopFactory;
+import org.apache.fop.apps.MimeConstants;
+
+public class BillingRaport {
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/BillingRaport.java]: please implement it as a spring service

File: src/main/java/com/example/event_manager/utils/BillingRaport.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+  private final BillingRaportSchema billingRaportSchema;
+  private String billingsXml;
+  static final String styleXslFile = "src/main/resources/static/billing/newStyle.xsl";
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/BillingRaport.java]: once you have it as a service, then inject this value as string or resource with @Value annotation https://www.baeldung.com/spring-value-annotation

File: src/main/java/com/example/event_manager/utils/newStyle.xsl [hunk 18]
```diff
@@ -0,0 +1,142 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (adamzareba) [inline src/main/java/com/example/event_manager/utils/newStyle.xsl]: move it to resource directory

**Pull request merged**